File: frontend/src/types/order.ts
/**
 * Order and Test Request Types
 */

// Import types from enums for local use
import type { TestStatus as TestStatusType } from '@/types/enums';
import type { OrderStatus as OrderStatusType } from '@/types/enums';
import type { PriorityLevel as PriorityLevelType } from '@/types/enums';
import type { PaymentStatus as PaymentStatusType } from '@/types/enums';
import type { ResultStatus as ResultStatusType } from '@/types/enums';
import type { ValidationDecision as ValidationDecisionType } from '@/types/enums';

// Re-export types (Single Source of Truth)
export type { TestStatus } from '@/types/enums';
export type { OrderStatus } from '@/types/enums';
export type { PriorityLevel } from '@/types/enums';
export type { PaymentStatus } from '@/types/enums';
export type { ResultStatus } from '@/types/enums';
export type { ValidationDecision } from '@/types/enums';

// Re-export the VALUES arrays for backwards compatibility
export { ORDER_STATUS_VALUES } from '@/types/enums';
export { PAYMENT_STATUS_VALUES } from '@/types/enums';

// Local type aliases for use in this file
type PriorityLevel = PriorityLevelType;
type ResultStatus = ResultStatusType;
type ValidationDecision = ValidationDecisionType;

/**
 * Type for result rejection during validation.
 * 're-test': Re-run test with same sample, creates new OrderTest
 * 're-collect': New sample required, triggers sample recollection
 * 'escalate': Escalate to supervisor when retest/recollect limits exceeded
 * 'authorize_retest': Escalation resolved with authorize re-test (history only)
 */
export type ResultRejectionType = 're-test' | 're-collect' | 'escalate' | 'authorize_retest';

/**
 * Record of a result rejection event during validation.
 * Stored in resultRejectionHistory array on OrderTest.
 * API may send either rejectionType/rejectionReason (camelCase) or type/reason.
 */
export interface ResultRejectionRecord {
  id?: number;
  resultId?: number;
  rejectedAt: string;
  rejectedBy: string;
  rejectionType?: ResultRejectionType;
  type?: ResultRejectionType;
  rejectionReason?: string;
  reason?: string;
  notes?: string;
}

/** Reads rejection type from either API shape (rejectionType or type). */
export function getResultRejectionType(
  record: ResultRejectionRecord
): ResultRejectionType | undefined {
  return record.rejectionType ?? record.type;
}

export interface TestResult {
  value: string | number;
  unit?: string;
  referenceRange?: string;
  status: ResultStatus;
}

export interface OrderTest {
  // Identity
  id?: number; // Integer ID, displayed as TST{id}
  testCode: string; // Links to Test catalog
  testName: string; // From API relationship
  sampleType: string; // From API relationship

  // Order-specific state
  status: TestStatusType;
  priceAtOrder: number; // Snapshot for billing

  // Sample linkage
  sampleId?: number; // Links to Sample (assigned after sample generation)

  // Results (order-specific)
  results: Record<string, TestResult> | null;
  resultEnteredAt?: string;
  enteredBy?: string;

  // Validation (order-specific)
  resultValidatedAt?: string;
  validatedBy?: string;
  validationNotes?: string;

  // Flags and notes
  flags?: string[];
  technicianNotes?: string;

  // Reflex/Repeat (order-specific)
  isReflexTest?: boolean;
  triggeredBy?: string;
  reflexRule?: string;
  isRepeatTest?: boolean;
  repeatReason?: string;
  originalTestId?: number;
  repeatNumber?: number;

  // Re-test tracking (for result validation rejection flow)
  isRetest?: boolean; // True if this is a retest of a rejected result
  retestOfTestId?: number; // Links to original OrderTest.id that was rejected
  retestNumber?: number; // 0 = original, 1 = 1st retest, etc.
  retestOrderTestId?: number; // Points to the new retest entry created after rejection

  // Result rejection history (for validation rejections)
  resultRejectionHistory?: ResultRejectionRecord[];

  // Critical values (order-specific)
  hasCriticalValues?: boolean;
  criticalNotificationSent?: boolean;
  criticalNotifiedAt?: string;
  criticalNotifiedTo?: string;
  criticalAcknowledgedAt?: string;

  // Metadata
  createdAt: string;
  updatedAt: string;
}

export interface Order {
  // Identity
  orderId: number; // Integer ID, displayed as ORD{id}
  patientId: number; // Links to Patient
  patientName: string; // From API relationship
  orderDate: string;

  // Tests
  tests: OrderTest[];

  // Pricing (snapshot at order time)
  totalPrice: number;
  paymentStatus: PaymentStatusType;
  overallStatus: OrderStatusType;

  // Scheduling
  scheduledCollectionTime?: string;

  // Instructions
  specialInstructions?: string[];
  patientPrepInstructions?: string;
  clinicalNotes?: string;
  referringPhysician?: string;
  priority: PriorityLevel;

  // Delivery tracking (optional - for detailed timeline)
  deliveredAt?: string; // When results were delivered
  deliveredBy?: string; // User who delivered results
  deliveryMethod?: 'email' | 'portal' | 'print' | 'sms';

  // Payment tracking (optional - for detailed timeline)
  paidAt?: string; // When payment was received
  paymentMethod?: string; // How payment was made

  // Metadata
  createdBy: string; // Backend returns string user ID
  createdAt: string;
  updatedAt: string;
}

/** Runtime-validated order shape (forms/mutations). */
export type { Order as ValidatedOrder, OrderTest as ValidatedOrderTest } from '@/features/orders/schemas/order.schema';

export interface ValidationRecord {
  validatedBy: string;
  validatedAt: string;
  decision: ValidationDecision;
  validationNotes?: string;
  criticalValuesNotified: boolean;
  rejectionReason?: string;
}

// Batch collection grouping
export interface BatchCollectionGroup {
  batchId: string;
  sampleType: string;
  orderIds: number[];
  patientCount: number;
  totalTests: number;
  priority: PriorityLevel;
  scheduledTime?: string;
  collectedBy?: string;
  collectedAt?: string;
  status: 'not-started' | 'in-progress' | 'completed';
}

File: backend/app/models/order.py
"""
Order and OrderTest Models - All fields use camelCase
"""
from sqlalchemy import Column, String, Float, DateTime, JSON, Enum, ForeignKey, Boolean, Integer
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base
from app.schemas.enums import OrderStatus, PaymentStatus, PriorityLevel, TestStatus


class Order(Base):
    __tablename__ = "orders"

    orderId = Column("order_id", Integer, primary_key=True, autoincrement=True, index=True)
    patientId = Column("patient_id", Integer, ForeignKey("patients.id"), nullable=False, index=True)
    orderDate = Column("order_date", DateTime(timezone=True), nullable=False)

    # Pricing
    totalPrice = Column("total_price", Float, nullable=False)
    paymentStatus = Column("payment_status", Enum(PaymentStatus), nullable=False, default=PaymentStatus.UNPAID)
    overallStatus = Column("overall_status", Enum(OrderStatus), nullable=False, default=OrderStatus.ORDERED)

    # Scheduling (optional - for future appointment integration)
    appointmentId = Column("appointment_id", Integer, nullable=True)
    scheduledCollectionTime = Column("scheduled_collection_time", DateTime(timezone=True), nullable=True)

    # Instructions
    specialInstructions = Column("special_instructions", JSON, nullable=True)  # Array of strings
    patientPrepInstructions = Column("patient_prep_instructions", String, nullable=True)
    clinicalNotes = Column("clinical_notes", String, nullable=True)
    referringPhysician = Column("referring_physician", String, nullable=True)
    priority = Column(Enum(PriorityLevel), nullable=False, default=PriorityLevel.ROUTINE)

    # Metadata
    createdBy = Column("created_by", String, nullable=False)
    createdAt = Column("created_at", DateTime(timezone=True), server_default=func.now())
    updatedAt = Column("updated_at", DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    patient = relationship("Patient", foreign_keys=[patientId])
    tests = relationship("OrderTest", back_populates="order", cascade="all, delete-orphan")

    @property
    def patientName(self) -> str:
        """Get patient full name from relationship"""
        return self.patient.fullName if self.patient else "Unknown"


class OrderTest(Base):
    __tablename__ = "order_tests"

    id = Column(Integer, primary_key=True, autoincrement=True, index=True)
    orderId = Column("order_id", Integer, ForeignKey("orders.order_id"), nullable=False, index=True)
    testCode = Column("test_code", String, ForeignKey("tests.code"), nullable=False, index=True)

    # Order-specific state
    status = Column(Enum(TestStatus), nullable=False, default=TestStatus.PENDING, index=True)
    priceAtOrder = Column("price_at_order", Float, nullable=False)  # Snapshot for billing

    # Sample linkage
    sampleId = Column("sample_id", Integer, nullable=True)  # Links to Sample

    # Results (JSON)
    results = Column(JSON, nullable=True)  # Record<string, TestResult>
    resultEnteredAt = Column("result_entered_at", DateTime(timezone=True), nullable=True)
    enteredBy = Column("entered_by", String, nullable=True)

    # Validation
    resultValidatedAt = Column("result_validated_at", DateTime(timezone=True), nullable=True)
    validatedBy = Column("validated_by", String, nullable=True)
    validationNotes = Column("validation_notes", String, nullable=True)

    # Flags and notes
    flags = Column(JSON, nullable=True)  # Array of strings
    technicianNotes = Column("technician_notes", String, nullable=True)

    # Reflex/Repeat
    isReflexTest = Column("is_reflex_test", Boolean, default=False)
    triggeredBy = Column("triggered_by", String, nullable=True)
    reflexRule = Column("reflex_rule", String, nullable=True)
    isRepeatTest = Column("is_repeat_test", Boolean, default=False)
    repeatReason = Column("repeat_reason", String, nullable=True)
    originalTestId = Column("original_test_id", Integer, nullable=True)
    repeatNumber = Column("repeat_number", Integer, nullable=True)

    # Re-test tracking (for result validation rejection flow)
    isRetest = Column("is_retest", Boolean, default=False)
    retestOfTestId = Column("retest_of_test_id", Integer, nullable=True)  # Links to original OrderTest.id that was rejected
    retestNumber = Column("retest_number", Integer, default=0)  # 0 = original, 1 = 1st retest, etc.
    retestOrderTestId = Column("retest_order_test_id", Integer, nullable=True)  # Points to the new retest entry created after rejection

    # Result rejection history (for validation rejections)
    resultRejectionHistory = Column("result_rejection_history", JSON, nullable=True, default=list)  # Array of ResultRejectionRecord

    # Critical values
    hasCriticalValues = Column("has_critical_values", Boolean, default=False)
    criticalNotificationSent = Column("critical_notification_sent", Boolean, default=False)
    criticalNotifiedAt = Column("critical_notified_at", DateTime(timezone=True), nullable=True)
    criticalNotifiedTo = Column("critical_notified_to", String, nullable=True)
    criticalAcknowledgedAt = Column("critical_acknowledged_at", DateTime(timezone=True), nullable=True)

    # Relationships
    order = relationship("Order", back_populates="tests")
    test = relationship("Test", foreign_keys=[testCode])

    @property
    def testName(self) -> str:
        """Get test display name from relationship"""
        return self.test.displayName if self.test else "Unknown"

    @property
    def sampleType(self) -> str:
        """Get test sample type from relationship"""
        return self.test.sampleType if self.test else "Unknown"
